Expose streaming API

adkelley opened the pull request:
# Expose Streaming API

## Motivation

Erlang's [inets](https://www.erlang.org/docs/26/apps/inets/) application provides a set of Internet-related services, including the HTTP client `httpc`. The `httpc` module gives users the option to stream the body of a 200 or 206 response either to the calling process or directly to a file.  

This pull request exposes that option to `gleam_httpc` while maintaining **full backward compatibility**. It supports the stated use case `stream:{self, once}`, (i.e., streaming chunks) represented by the custom type `Stream`:

```gleam
type Destination {
  Self
}

type Mode {
  Once
}

type ErlOption {
  ...
  Stream(#(Destination, Mode))
}

```
None of the other streaming APIs from the Erlang httpc module is supported (e.g., stream:{filename}).

## Implementation Details

Several functions that have synchronous equivalents were added to support streaming, including:

* `dispatch_stream_bits(config: Configuration, req: Request(BitArray))` - similar to `dispatch_bits` function,  send a HTTP stream request of binary data
* `dispatch_stream_request(config: Configuration, req: Request(String))` - similar to the`dispatch` function, send a HTTP stream request of unicode data
*  `send_stream_request(req: Request(String)` - similar to the `send` function, send an HTTP stream request of Unicode data using the default configuration.

The function `select_stream_messages` configures a selector to receive stream messages.   Finally, the function `receive_next_stream_message(pid: Pid)` triggers the next streaming message to be sent to the calling process designated by `pid`. 
  
## Testing

Added two new unit tests, `send_stream_request_test`, and `dispatch_stream_request_test` using the [postman streamed response api](https://www.postman.com/postman/postman-public-workspace/request/8sqcxxz/streamed-response)

## Documentation

* Updated README.md with a `stream:{self, once}` example
* Added inline documentation to new public functions 

### Related Issues

* Resolves Issue #31

adkelley committed:
Expose streaming API

Erlang's httpc module supports streaming the body of 200 or 206
responses to the calling process or to a file.  These changes expose
this API.

adkelley committed:
Expose streaming API - stream:{self, once}

Erlang's httpc module supports streaming the body of 200 or 206
responses to the calling process or to a file.  These changes expose the
stream:{self, once} API, the stated use case.

Resolves Issue Expose streaming API #31

adkelley committed:
Expose streaming API - stream:{self, once}

Erlang's httpc module supports streaming the body of 200 or 206
responses to the calling process or to a file.  These changes expose the
stream:{self, once} API, the stated use case.

Resolves Issue Expose streaming API #31

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +136,161 @@ pub fn send_bits(
+/// `List(#(Charlist, Charlist))` and `Pid` respectively for further
+/// processing by the caller.
+@external(erlang, "gleam_stdlib", "identity")
+pub fn dynamic_to_headers_pid(s: Dynamic) -> #(List(#(Charlist, Charlist)), Pid)
```
This function is not safe, remove it.

adkelley replied:
Removed.  See my response to your comment suggesting that we restrict the api to stream:{self, once}

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +136,161 @@ pub fn send_bits(
+/// data `#(stream_start, headers)` This convenience function that converts
+/// `headers` to `List(#(Charlist, Charlist))` for further processing by the caller.
+@external(erlang, "gleam_stdlib", "identity")
+pub fn dynamic_to_headers(s: Dynamic) -> List(#(Charlist, Charlist))
```
This function is not safe, remove it.

adkelley replied:
Removed.  See my response to your comment suggesting that we restrict the api to stream:{self, once}

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +136,161 @@ pub fn send_bits(
+/// is the `Dynamic` data `#(stream, chunk)` This is a convenience function that converts
+/// `chunk` to a `BitArray` for further processing by the caller.
+@external(erlang, "gleam_stdlib", "identity")
+pub fn dynamic_to_bit_array(s: Dynamic) -> BitArray
```
This function is not safe, remove it.

adkelley replied:
Removed.  See my response to your comment suggesting that we restrict the api to stream:{self, once}

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +136,161 @@ pub fn send_bits(
+@external(erlang, "gleam_stdlib", "identity")
+pub fn dynamic_to_response(
+  s: Dynamic,
+) -> #(#(Charlist, Int, Charlist), List(#(Charlist, Charlist)), BitArray)
```
This function is not safe, remove it.

adkelley replied:
Removed.  See my response to your comment suggesting that we restrict the api to stream:{self, once}

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +136,161 @@ pub fn send_bits(
   |> dispatch_bits(req)
 }
 
+// region:    --- asynchronous HTTP request
```
Remove this region comment please 🙏

adkelley replied:
Removed

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +136,161 @@ pub fn send_bits(
+pub fn async_receive(
+  req_id: Reference,
+  timeout: Int,
+) -> Result(#(Atom, Dynamic), HttpError) {
```
This type is under-specified! Rather than the super general type of `#(Atom, Dynamic)` this should be a type that precisely specifies the shape of the data. That makes it easier and more convenient to use, and the type system will make mistakes impossible.

adkelley replied:
I removed this function, replacing it with `receive_stream_start`, receive_stream_chunk`, and `receive_stream_end`.  This enables deterministic return type signatures and eliminates  the necessity for `#(Atom, Dynamic)`

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -166,6 +362,12 @@ pub opaque type Configuration {
+    /// 
+    /// Default is `StreamNone`, which buffers and send the entire response
+    /// in one final message.
+    stream: Stream,
```
I don't think this should be in the builder record. The various send+dispatch+async functions here all expect some specific streaming configuration, so by being able to set it here and also in those functions makes it possible to represent invalid states.

adkelley replied:
Yes.  I removed it by limiting the API to stream:{self, once}

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +136,161 @@ pub fn send_bits(
+/// using the convenience functions `httpc.dynamic_to_header_pid`,
+/// `httpc.dynamic_to_header`, and `httpc.dynamic_to_bit_array`
+/// TODO: Refine Error
+pub fn async_receive(
```
In Gleam we receive messages using the `Subject` and `Selector` types from `gleam_erlang`. If this package invents a new way to do message receiving it means it will be incompatible with all other Gleam OTP code, and so you won't be able to use it within actors, etc. It also means one cannot wait on multiple HTTP requests at the same time.

adkelley replied:
Interesting.  I've eliminated all `dynamic_to_xx` convenience functions as a consequence of limiting the API to stream:{self, once}.  Please review my changes to `gleam_httpc_ffi.erl` and share your thoughts and suggestions on alternative ways to support these messages, while aligning with your point above.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +136,161 @@ pub fn send_bits(
+@external(erlang, "gleam_httpc_ffi", "stream_next")
+fn stream_next(pid: Pid) -> Result(Nil, Nil)
+
+// TODO: refine error type
```
Is this still to do?

adkelley replied:
I copied your comment `TODO: refine error type` when documenting my asynchronous equivalents to your synchronous functions (e.g., send_bits).  If you don't plan on further refining the error type, then we can remove it.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -41,8 +44,27 @@ type BodyFormat {
+  /// before receiving the next chunk.
+  StreamSelfOnce
+  /// Stream response body directly into a file.
+  StreamFile(String)
```
There's 3 new streaming APIs being added here, do we have sufficient motivation to add them all? The stream-self one is rather dangerous as it can overwhelm the process with messages.

It seems to me there's a choice between streaming chunks as they arrive, streaming to a file, and asynchronously receiving the response later. The linked issue reads to me as if it would require streaming chunks, and I don't think we have a stated use case for the other two yet.

adkelley replied:
Yes, you make a compelling argument for limiting the API to stream:{self, once}.  This is my use case, which was the impetus for this PR.  By limiting the API to stream:{self, once}, I eliminated the requirement for the Dynamic type. Consequently, the return type signatures from the three streaming messages `stream_start`, `stream`, and `stream_end` are now well-typed.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -41,8 +44,27 @@ type BodyFormat {
   BodyFormat(BodyFormat)
   SocketOpts(List(SocketOpt))
+  Sync(Bool)
+  Stream(Dynamic)
```
Best to specify this precisely, so the type system can help prevent mistakes.

adkelley replied:
Yes, after limiting the API to `stream:{self, once}`.  its now well typed  -`Stream(#(Atom, Atom))`

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +136,161 @@ pub fn send_bits(
+pub fn async_stream_next(pid: Pid) -> Result(Nil, HttpError) {
+  use _next <- result.try(
+    // TODO return reason somehow? {http, {ReqId, {error, Reason}}}
+    stream_next(pid) |> result.replace_error(InvalidUtf8Response),
```
This doesn't seem like the right error value?

adkelley replied:
Yes, you're right.  Actually, with Erlang's `httpc:stream_next/1`, if the Pid is not a valid HTTP client handler (i.e., not managing a live streaming request), then the call will silently fail with no effect. It will still return ok, but no message will be sent, and no chunk will arrive.   I decided to handle this by returning `Result(Nil, Nil)`, regardless of the impossibility of an error.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +136,161 @@ pub fn send_bits(
+    |> result.map_error(with: fn(reason) {
+      case atom.to_string(reason) {
+        "timeout" -> ResponseTimeout
+        _ -> InvalidUtf8Response
```
How come this isn't done with the existing error conversion code?

adkelley replied:
My mistake, I'm now using `result.map_error(normalize)` in all cases.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -224,6 +454,31 @@ pub fn dispatch(
+  let request = request.map(request, bit_array.from_string)
+  use request_id <- result.try(
+    async_dispatch_bits(config, request)
+    // TODO Is this the right error? Should I create a unique streaming error?
```
What are the errors that can be returned by `async_dispatch_bits`?

adkelley replied:
Great question.  There is no comprehensive, officially documented list of all possible Reason values for {http, {ReqId, {error, Reason}}} in Erlang’s httpc.  Moreover, the Reason in {error, Reason} can come from multiple layers of the inet stack.  Consequently, I'm using your `result.map_error(normalize)`, which I believe comprehensively covers the most likely ones (e.g., ResponseTimeOut).

lpil reviewed (commented):
Thank you! I've left a bunch of notes inline, let me know if any of it is unclear.

adkelley committed:
Restrict api to stream:{self, once}

Streaming chunks is the stated use case. By limiting support to this API
(i.e, stream:self, once} it removes considerable complexity and enables
deterministic API responses that can be correctly typed in gleam

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -41,6 +44,9 @@ type BodyFormat {
   SocketOpts(List(SocketOpt))
+  Sync(Bool)
+  // {self, once}
+  Stream(#(Atom, Atom))
```
Use custom types rather than atoms please, to prevent invalid data from being able to be passed to the functions 🙏

adkelley replied:
```gleam
type Destination {
  Self
}

type Mode {
  Once
}

type ErlOption {
  ...
  Stream(#(Destination, Mode))
}
```

adkelley replied:
Yes, [https://hexdocs.pm/mug/mug.html#select_tcp_messages](https://hexdocs.pm/mug/mug.html#select_tcp_messages)  was a helpful example.  In my case, the calling process receives the selector using `  let selector = httpc.initialize_stream_selector()` and can issue `process.selector_receive(selector, timeout)` calls to get the next stream message.

lpil reviewed (commented):
This is looking a lot nicer, thank you!

It still needs to be changed to use selectors rather than a custom non-composable API for receiving messages. Check out Mug's <https://hexdocs.pm/mug/mug.html#receive_next_packet_as_message> and <https://hexdocs.pm/mug/mug.html#select_tcp_messages> for how this can be done.

The new message types are not very clear due to using tuples instead of a descriptive custom type, and the use of charlists makes it difficult for the programmer to use as they have to know how to convert them to strings. Make a custom type for the response messages please, one designed with the Gleam programmer in mind rather than being more faithful to the Erlang APIs are we building on top of 🙏 

Thanks again!

adkelley committed:
Merge branch 'self_once'

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -59,6 +73,21 @@ type ErlVerifyOption {
+  List(#(String, String))
+
+type RequestId =
+  Reference
```
The indirection this alias adds makes the code more confusing, please remove this.

If you wish to make a distinct request id type it would be done like so:

```gleam
pub opaque type RequestIdentifier {
  RequestIdentifier(reference: Reference)
}
```

Or if we don't in Gleam need to know about the internal Erlang structure being a reference then an external type may be most suited.

```gleam
pub type RequestIdentifier
```
This could be a good fit here as it adds type safety, has no runtime overhead, and there's no need to manually wrap and unwrap it.

adkelley replied:
Nice!  Yes, neither Gleam nor the caller needs to know that it's an Erlang reference type.  Just that they may need to filter on different id numbers when making  I went with `pub type RequestIdentifier` for type safety reasons

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -59,6 +73,21 @@ type ErlVerifyOption {
+pub type HttpSocket
+
+type Headers =
+  List(#(String, String))
```
Remove this alias please.

adkelley replied:
Removed

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +143,105 @@ pub fn send_bits(
+pub fn async_send_bits(req: Request(BitArray)) -> Result(Reference, HttpError) {
+  configure()
+  |> async_dispatch_bits(req)
+}
```
Let's have just the dispatch version that takes config.

adkelley replied:
Removed

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +143,105 @@ pub fn send_bits(
+
+/// Send an asynchronous HTTP request of binary data
+/// 
+pub fn async_dispatch_bits(
```
I think the name and documentation comments here are unclear, I can't tell from them what this function does. A more descriptive name would be fab, and documentation that explains what this function does and what other functions would be used in order to receive the body would be fab. Examples too would likely be needed for a new Gleam/OTP programmer to understand how to use it.

The word "async" is possibly misleading. Languages that use the term "async" (JS, Python, Rust, C#, etc) tend use it to mean it uses coroutines to multiplex tasks onto a single OS thread, instead of the blocking it. This is actually what the existing "send" and "dispatch" functions do (and all code on the BEAM for that matter), so I think to use that highly loaded term here would give the wrong impression and result in people going down the wrong path when they want to make simple HTTP requests.

adkelley replied:
Renamed to the public api functions to `send_stream_request ',  and ' dispatch_stream_request`.  Documentation added.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +143,105 @@ pub fn send_bits(
+/// If you wish to only handle stream messages from one process, then use one
+/// process per asychronous HTTP request. 
+///
+pub fn initialize_stream_selector() -> process.Selector(StreamMessage) {
```
Use the same design as found in the other packages please, so take the selector as an argument, take a mapping function as an argument, and use the naming convention they use. The TCP library is a good reference here.

If you cannot use an existing selector then you would not be able to use it with code that already has a selector (so most actor based code).

if you do not have a mapping function then you can't produce selectors of different types, which means you cannot easily select other types of messages with the selector.

And the naming convention is just to be consistent and to make the API easy to approach for people who have learnt about how to use selectors already.

adkelley replied:
Yes, agreed.  Revised.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -224,7 +368,26 @@ pub fn dispatch(
+  let request = request.map(request, bit_array.from_string)
+  use request_id <- result.try(async_dispatch_bits(config, request))
+  Ok(request_id)
+}
```
This is named as if it works with strings, but the body type returned is a bit array. Maybe we don't need this version and just working always with bit arrays is fine.

lpil replied:
Let's always work with bits and not have a string specialised version, seeing as it is not possible to ensure that the response is a string, as the other string versions do.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +143,105 @@ pub fn send_bits(
+pub fn stream_next(pid: process.Pid) -> Result(Nil, Nil)
+
+@external(erlang, "gleam_httpc_ffi", "coerce_stream_message")
+fn unsafe_decode(msg: Dynamic) -> StreamMessage
```
More descriptive name for this please 🙏 

Looks like from the Erlang it is safe (it doesn't do an unchecked cast), but it asserts the format.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +143,105 @@ pub fn send_bits(
+
+fn map_stream_message(mapper: fn(StreamMessage) -> t) -> fn(Dynamic) -> t {
+  fn(message) { mapper(unsafe_decode(message)) }
+}
```
Inline the anonymous function into the select function please, for clarity

adkelley replied:
Done

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +143,105 @@ pub fn send_bits(
+  selector: process.Selector(t),
+  mapper: fn(StreamMessage) -> t,
+) -> process.Selector(t) {
+  let http = atom.create("http")
```
Could use `Http` from `gleam/http` here

adkelley replied:
Revised to `let http = atom.create(http.scheme_to_string(http.Http))`

lpil commented on src/gleam_httpc_ffi.erl:
```diff
@@ -1,5 +1,24 @@
+ case httpc:stream_next(HandlerPid) of
+   ok -> {ok, nil};
+   _  -> {error, nil}
+ end.
```
4 space indentation please 🙏 

That said, this function never fails so this wrapper can be removed. Have it return `Nil`, and give it a descriptive name, similar to how the Mug package does it.

adkelley replied:
Yes.  This now returns Nil, and is renamed to `receive_next_stream_message`.  Consequently, the case statement was removed.

lpil commented on src/gleam_httpc_ffi.erl:
```diff
@@ -1,5 +1,24 @@
+   _  -> {error, nil}
+ end.
+
+coerce_stream_message({http, {ReqId, stream_start, Headers, Pid}}) -> {stream_start, ReqId, Headers, Pid};
```
Could you wrap these long lines please 🙏 

I think the headers will need to be mapped from charlists to binary strings?

A guard for the body to make sure it is a bit array would be fab.

What's the difference between the headers on the stream start and end?

adkelley replied:
Added a guard for the body.  I chose to return raw stream messages, then use your `string_header` function to convert the headers from `List(#(Charlist, Charlist)) to `List(#(String, String))`

lpil commented on README.md:
````diff
@@ -8,6 +8,9 @@ Bindings to Erlang's built in HTTP client, `httpc`.
 gleam add gleam_httpc@5
 ```
+
+## Synchronous request
````
I think this going to be misleading for people unfamiliar with the BEAM- it is an async request. See my comment above about the term "async".

adkelley replied:
Yes, revised to make it easy to understand.  I also documented what stream:{self, once} is.

lpil reviewed (commented):
Lovely work! This is def getting where we want it to be. I've left more notes inline, thank you

adkelley committed:
Merge branch 'self_once'

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -59,6 +72,24 @@ type ErlVerifyOption {
   VerifyNone
 }
 
+pub type HttpSocket
```
Is this used anywhere?

adkelley replied:
Good catch  - removed.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -59,6 +72,24 @@ type ErlVerifyOption {
 
+pub type HttpSocket
+
+pub type RequestIdentifier
```
What's this? As it's part of the public API it'll need to be documented

adkelley replied:
Not part of private or public API - removed

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -59,6 +72,24 @@ type ErlVerifyOption {
+  RawStreamChunk(RequestIdentifier, BitArray)
+  RawStreamEnd(RequestIdentifier, List(#(Charlist, Charlist)))
+  RawStreamError(RequestIdentifier, HttpError)
+}
```
This looks like it is an implementation detail and should not be in part of the public interface?

adkelley replied:
Yes, it's an implementation detail.  I've refactored the code to make it private, and, as a consequence, I was able to simplify the configuration of the selector that receives stream messages (See  `pub fn select_stream_messages() -> process.Selector(StreamMessage)`)

lpil replied:
How come we construct values of this type and then immediately convert to another type? Why not construct that final type and skip this intermediate one?

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -59,6 +72,24 @@ type ErlVerifyOption {
+  StreamChunk(RequestIdentifier, BitArray)
+  StreamEnd(RequestIdentifier, List(#(String, String)))
+  StreamError(RequestIdentifier, HttpError)
+}
```
Could you document this type and add labels please 🙏

adkelley replied:
Done

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +145,124 @@ pub fn send_bits(
 }
 
+/// Send a HTTP stream request of binary data
+/// 
```
Could you finish this documentation please 🙏

adkelley replied:
Done

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +145,124 @@ pub fn send_bits(
+  Ok(request_id)
+}
+
+/// Triggers the next asynchronous streaming message to be sent to the calling process
```
Wrap before 80 columns please 🙏 

Could we use a name more like the name in the other library please

lpil commented on src/gleam/httpc.gleam:
````diff
@@ -98,6 +145,124 @@ pub fn send_bits(
+/// process.new_selector() |> select_stream_messages(raw_stream_mapper())
+/// ```
+///
+pub fn raw_stream_mapper() -> fn(RawStreamMessage) -> StreamMessage {
````
This looks like part of the internal implementation and shouldn't be part of the public API 🙏

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -209,7 +374,7 @@ pub fn timeout(config: Configuration, timeout: Int) -> Configuration {
 }
 
-/// Send a HTTP request of unicode data.
+/// Send a synchronus HTTP request of unicode data.
```
Remove this please 🙏

adkelley replied:
Done

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -224,7 +389,58 @@ pub fn dispatch(
+  configure()
+  |> dispatch_stream_request(req)
+}
+
```
Let's have just the dispatch function 🙏

adkelley replied:
Can you explain more about this?  Are you suggesting that I remove `pub fn send_stream_request(req: Request(String)) -> Result(RequestIdentifier, HttpError)`, leaving `dispatch_stream_request` whenever streaming?

lpil replied:
Yes please

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -224,7 +389,58 @@ pub fn dispatch(
+/// 
+/// With the exception of timeout errors, all other errors will be delivered via:
+/// `StreamError(RequestIdentifier, HttpError)`.
+///
```
This is a complex API, so folks will need examples to understand how to use it. Could you add these please 🙏

adkelley replied:
I've added a thorough example to the README

adkelley replied:
I also added a simpler example to the documentation above `dispatch_stream_request` and `send_stream_request` functions.

lpil commented on test/gleam_httpc_test.gleam:
```diff
@@ -161,3 +162,61 @@ pub fn timeout_error_test() {
+    process.selector_receive(selector, 1000)
+  assert request_id_ == request_id
+
+  let _nil = httpc.receive_next_stream_message(pid)
```
Remove the `let nil_ =` prefixes please, including from the documentation

adkelley replied:
Done.

lpil commented on test/gleam_httpc_test.gleam:
```diff
@@ -161,3 +162,61 @@ pub fn timeout_error_test() {
+  let assert Ok(httpc.StreamEnd(request_id_, _headers)) =
+    process.selector_receive(selector, 1000)
+  assert request_id_ == request_id
+}
```
Can we have a test for the error case please

adkelley replied:
Added a test that closes the connection mid-response, triggering stream error case

lpil reviewed (commented):
Thank you! Nearly there, left some API related notes inline

adkelley committed:
Refactor streaming code, tests, and docs

* Updated README with a streaming example

* Added error case to streaming tests

* Removed send_stream from httpc and improved function documentation

lpil commented on README.md:
```diff
@@ -37,6 +38,66 @@ pub fn send_request() {
 
+## Http streaming requests
+
+`httpc` supports `stream:{self, once}` mode, which is a **pull-based** approach for
```
What's this `stream:{self, once}` syntax? I don't recognise it

adkelley replied:
Yes, that's Erlang speak :-) Replaced with `Stream(#(Self, Once))

lpil commented on README.md:
```diff
@@ -37,6 +38,66 @@ pub fn send_request() {
+
+/// Receive a streamed response from Postman Echo. The number of
+/// stream chunks we receive is 5, as we specfied by the endpoint
+///
```
Remove the reference to postman please

adkelley replied:
Removed

lpil commented on README.md:
```diff
@@ -37,6 +38,66 @@ pub fn send_request() {
+  // Configure the selector
+  let selector = httpc.select_stream_messages()
+  let assert Ok(chunks) =
+    loop(request_id, selector, process.self(), bit_array.from_string(""))
```
Do not use raw looping please, we don't want people to do that when they should be using an actor

adkelley replied:
Revised to use the Actor model.  This required a good deal of thought, but it certainly helped me to better understand your approach to OTP.

lpil commented on test/gleam_httpc_test.gleam:
```diff
@@ -161,3 +162,74 @@ pub fn timeout_error_test() {
+  let selector = httpc.select_stream_messages()
+  let assert Ok(_request_id) = httpc.send_stream_request(req)
+  let assert Ok(httpc.StreamError(_request_id, httpc.SocketClosedRemotely)) =
+    process.selector_receive(selector, 1000)
```
Use `assert process.selector_receive(selector, 1000) == Ok(httpc.StreamError(request_id, httpc.SocketClosedRemotely))` please 🙏

adkelley replied:
Done

lpil commented on src/gleam_httpc_ffi.erl:
```diff
@@ -1,5 +1,22 @@
+coerce_stream_message({http, {ReqId, stream_start, Headers, Pid}}) -> {raw_stream_start, ReqId, Headers, Pid};
+coerce_stream_message({http, {ReqId, stream, BinBodyPart}}) when is_binary(BinBodyPart) -> {raw_stream_chunk, ReqId, BinBodyPart};
+coerce_stream_message({http, {ReqId, stream_end, Headers}}) -> {raw_stream_end, ReqId, Headers};
+coerce_stream_message({http, {ReqId, {error, Reason}}}) -> {raw_stream_error, ReqId, normalise_error(Reason)}. 
```
Wrap these long lines before 80 columns please

adkelley replied:
Done

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -224,7 +400,137 @@ pub fn dispatch(
+/// delivered to the caller.
+/// 
+/// With the exception of timeout errors, all other errors will be delivered
+/// via: `StreamError(RequestIdentifier, HttpError)`.  
```
Is it possible to get multiple errors from a single stream? What happens after an error is received for a request?

adkelley replied:
StreamErrors propagate back to the user, leaving the user to decide whether to continue, terminate, or reinitialize the stream request.  See my example in the README.md, where I use `actor.stop_abnormal("Http error")` to shut down the streaming process whenever the first StreamError message is received.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -224,7 +400,137 @@ pub fn dispatch(
+/// This function supports only the `stream: {self, once}` mode from `httpc`,
+/// which is a **pull-based** streaming approach. In this mode, the caller must
+/// explicitly request the next stream message using
+/// `receive_next_stream_message`.
```
Same question here RE that syntax.

Please reword these to explain the functionality rather than making a comparison to the Erlang library, and to remove the emphasis this the "pull based" concept. I think we can omit that phrase entirely as it seems the entire response flow is pull based, but really it's push with only the very last part being pull based.

adkelley replied:
Yes, replaced this Erlang syntax with the corresponding Gleam syntax.  I omitted the phrase entirely.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -177,7 +353,7 @@ pub opaque type Configuration {
 /// - The timeout for the response to be received is 30 seconds from when the
 ///   request is sent.
-///
+/// 
```
```suggestion
///
```

adkelley replied:
Removed the extra space

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +189,90 @@ pub fn send_bits(
+  }
+  process.new_selector()
+  |> process.select_record(http, 1, map_stream_message(mapper))
+}
```
Have this take the selector as an argument, to match all other selecting code. Forcing a new selector means the programmer cannot use this library any time that they need to use an existing selector.

adkelley replied:
Yes, your mug example (i.e. ` mug.select_tcp_messages`) was very helpful here.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -224,7 +400,137 @@ pub fn dispatch(
+///    let selector = httpc.select_stream_messages()
+///    let assert Ok(httpc.StreamStart(_request_id_, _headers, handler_pid)) =
+///       process.selector_receive(selector, 1000)
+///    httpc.receive_next_stream_message(handler_pid)
```
I think it's rather confusing that this is getting the pid from here and not using `process.self()`. It seems to imply that this could be changed, or that the message should not go to this pid

adkelley replied:
httpc returns a unique process ID in the `stream_start` message tuple, which we use to trigger the next streaming message, and subsequent ones.  I've changed this variable name to `stream_pid`, because I think it better describes it's purpose.  The mental model I'm using: Mental model to keep things straight:

request_id identifies the request
streami_pid identifies the flow-control gate for the response body

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -98,6 +189,90 @@ pub fn send_bits(
+}
+
+/// Triggers the next streaming message to be sent to the calling process
+/// designated by `pid`. 
```
Is it to the calling process or the process that the pid is for?

adkelley replied:
Yes, this is confusing.  Revised to  
/// Triggers the next streaming message to be sent to the calling process.
/// Use the `stream_pid` argument returned by the `RawStreamStart` message.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -59,6 +74,66 @@ type ErlVerifyOption {
+/// 
+pub type StreamMessage {
+  /// Sent exactly once when the server response begins. The returned `pid`
+  /// identifies the process and must be passed to `httpc:stream_next`
```
No Erlang syntax or explaining Erlang APIs in Gleam documentation please.

adkelley replied:
Revised to refer to the Gleam function

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -59,6 +74,66 @@ type ErlVerifyOption {
+  )
+  /// Sent for every chunk of response data that the worker emits. Each chunk
+  /// must be explicitly requested by calling `httpc:stream_next/1` with the pid
+  /// supplied in `StreamStart`.
```
Remove the documentation of Erlang internals please. This is documentation for users, not maintainers.

lpil commented on src/gleam/httpc.gleam:
```diff
@@ -59,6 +74,66 @@ type ErlVerifyOption {
+pub type StreamMessage {
+  /// Sent exactly once when the server response begins. The returned `pid`
+  /// identifies the process and must be passed to `httpc:stream_next`
+  /// whenever you are ready for the next chunk.
```
What is this process? Is it some internal detail of Erlang's HTTPC? Is it a process the user creates? What happens if we pass a different pid to the stream function?

adkelley replied:
Good question!  According to the API documentation, you're supposed to use this pid in `stream_next_message`. However, I tested a different pid (process.self()) in an example, and it still worked correctly.  I suggest we stick with the documented behavior (See https://www.erlang.org/doc/apps/inets/httpc.html#request/5)

lpil reviewed (commented):
Oh sorry! I wrote some review comments and then didn't press submit for some reason. Sorry about that

Munksgaard commented:
What's the status on this PR? Streaming seems like something that'd be nice to have :-)

jtdowney commented on src/gleam_httpc_ffi.erl:
```diff
@@ -1,5 +1,27 @@
+ httpc:stream_next(HandlerPid),
+ nil.
+
+coerce_stream_message({http, {ReqId, stream_start, Headers, Pid}}) ->
```
I was trying out this branch with the Ollama API because I'm excited to see it get landed, and I ran into a bug.

Due to an issue in my environment, the remote server returned an error message as a complete HTTP response (with a content-length and not chunked) instead of a streaming response. `httpc` sends that as a `{http, {RequestId, Result}}` message. Since the message wasn't matched here, it failed when trying to coerce it into a streaming message.

```
runtime error: Erlang error

No Erlang function clause matched the arguments it was called with.

stacktrace:
  gleam_httpc_ffi.coerce_stream_message gleam_httpc_ffi.erl:13
  gleam/httpc.-select_stream_messages/2-anonymous-0- src/gleam/httpc.gleam:293
  gleam_erlang_ffi.select gleam_erlang_ffi.erl:64
  ollama_streaming.main dev/ollama_streaming.gleam:67
```
